mypy/semanal.py

```diff
@@ -6408,9 +6408,31 @@ def lookup_fully_qualified_or_none(self, fullname: str) -> SymbolTableNode | Non
         # TODO: support nested classes (but consider performance impact,
         #       we might keep the module level only lookup for thing like 'builtins.int').
         assert "." in fullname
+
         module, name = fullname.rsplit(".", maxsplit=1)
+
+        # The reason for this is that we want to be able to handle cases such as importlib.machinery
         if module not in self.modules:
-            return None
+            # Check if there's nested module A.B.C
+            splitted = fullname.rsplit(".")
+            module, names = splitted[0], splitted[1:]
```

Comment:
I think this logic is not quite right for packages. We should find the longest prefix of fullname that is a module (it could be e.g. `a.b` for `a.b.C.D`), and then look up through TypeInfos based on that.

Also add a test case for a case like `a.b.C.D` where `a.b` is a package.

Reply:
Sorry about that! I'll take a look as soon as possible :)